lib/jnpr/junos/factcache.py

```diff
@@ -0,0 +1,280 @@
+                                            new_key))
+                    else:
+                        # Cache the returned fact
+                        self._cache[new_key] = new_facts[new_key]
+            finally:
+                # Always pop the current callback from _call_stack,
+                # regardless of whether or not an exception was raised.
+                self._call_stack.pop()
+        if key in self._cache:
+            # key fact is cached. Return it.
+            if self._device._fact_style == 'both':
+                # Compare old and new-style values.
+                if key in self._device._ofacts:
+                    # Skip key comparisons for certain keys.
+                    #
+                    # The old facts gathering code has an up_time key.
+                    # The new facts gathering code maintains this key for RE0
+                    # and RE1 facts, but it's not comparable (because it
+                    # depends on when the fact was gathered and is therefore
+                    # not really a "fact".) The new re_info fact omits the
+                    # up_time field for this reason.
+                    #
+                    # The old facts gathering code didn't return a correct
+                    # value for the master fact when the system was a VC.
+                    # The new fact gathering code still returns the master fact
+                    # but returns a correct value for VCs. It also returns a
+                    # new re_master fact which is much more useful.
+                    if key not in ['RE0', 'RE1', 'master']:
+                        if self._cache[key] != self._device._ofacts[key]:
+                            warnings.warn('New and old-style facts do not '
+                                          'match for the %s fact.\n'
+                                          '    New-style value: %s\n'
+                                          '    Old-style value: %s\n' %
+                                          (key,
+                                           self._cache[key],
+                                           self._device._ofacts[key]),
+                                          RuntimeWarning)
+            return self._cache[key]
+        else:
+            # key fact was not returned by callback
+            raise RuntimeError("The %s module claims to provide the %s "
+                               "fact, but failed to return it. Please report "
+                               "this error." %
+                               (self._callbacks[key].__module__, key))
+
+    def __delitem__(self, key):
+        """
+        Facts are read-only. Don't allow deleting an item.
+        """
+        raise RuntimeError("facts are read-only!")
+
+    def __setitem__(self, key, value):
+        """
+        Facts are read-only. Don't allow setting an item.
+        """
+        raise RuntimeError("facts are read-only!")
```

Comment:
I have not yet changed this behavior. Since these are supposed to be "facts" about the device it seems inappropriate to allow users to set a fact to an arbitrary value. The `facts_refresh()` method already provides a way to force the refresh of specific facts.

Can you please explain the use case for setting "facts" to a user-defined value?